rocmPackages: extend ISA compatibility

GZGavinZhao opened the pull request:
- **rocmPackages: extend rocm-runtime ISA compatibility**
- **rocmPackages: extend clr ISA compatibility**
- **rocmPackages: extend tensile ISA compatibility**
- **rocmPackages: extend rocblas ISA compatibility**
- **rocmPackages: extend miopen ISA compatibility**

## Description of changes

This will eliminate the need for users to use `HSA_OVERRIDE_GFX_VERSION` to emulate their GPUs as a supported GPU model.

For more info, see [this](https://lists.debian.org/debian-ai/2024/02/msg00178.html) mailing thread.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

GZGavinZhao committed:
rocmPackages: extend rocm-runtime ISA compatibility

Signed-off-by: Gavin Zhao <git@gzgz.dev>

GZGavinZhao committed:
rocmPackages: extend clr ISA compatibility

Signed-off-by: Gavin Zhao <git@gzgz.dev>

GZGavinZhao committed:
rocmPackages: extend tensile ISA compatibility

Signed-off-by: Gavin Zhao <git@gzgz.dev>

GZGavinZhao committed:
rocmPackages: extend rocblas ISA compatibility

Signed-off-by: Gavin Zhao <git@gzgz.dev>

GZGavinZhao committed:
rocmPackages: extend miopen ISA compatibility

Signed-off-by: Gavin Zhao <git@gzgz.dev>

mschwaig commented:
I think having that artificial barrier removed is a great change.

I just want to give you a heads up after quickly looking over that linked email:

It seems those patches are originally provided as git commits. Unless they need to be adapted somehow form the original commits, it would be better to integrate them by using `fetchpatch` instead of checking them in, like you can see for example in this change : https://github.com/NixOS/nixpkgs/commit/9a4f48bb251a2275293a85611a503b46bbdcf9cb

GZGavinZhao commented:
@mschwaig Updated. Thank you for your comment!

GZGavinZhao commented:
I will leave `nixpkgs-review` running through the night, and if nothing fails I'll mark squash the commits and mark this as ready for review tomorrow.

GZGavinZhao commented:
Well I guess that had to be 2 nights because it ran out of space building `rocblas`... Going to finish the rest tonight and hopefully I can update the PR tomorrow.

GZGavinZhao committed:
rocmPackages: respect NIX_BUILD_CORES and cleanup CMake options in rocblas

Signed-off-by: Gavin Zhao <git@gzgz.dev>

mschwaig commented:
Going by the output of `nixpkgs-review` it seems like `rocblas` is still not building yet.

Result of `nixpkgs-review pr 298388` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>14 packages marked as broken and skipped:</summary>
  <ul>
    <li>rocmPackages.llvm.flang</li>
    <li>rocmPackages.llvm.flang.doc</li>
    <li>rocmPackages.llvm.flang.info</li>
    <li>rocmPackages.llvm.flang.man</li>
    <li>rocmPackages.migraphx</li>
    <li>rocmPackages.rdc</li>
    <li>rocmPackages.rdc.doc</li>
    <li>rocmPackages_6.llvm.flang</li>
    <li>rocmPackages_6.llvm.flang.doc</li>
    <li>rocmPackages_6.llvm.flang.info</li>
    <li>rocmPackages_6.llvm.flang.man</li>
    <li>rocmPackages_6.migraphx</li>
    <li>rocmPackages_6.rdc</li>
    <li>rocmPackages_6.rdc.doc</li>
  </ul>
</details>
<details>
  <summary>6 packages failed to build:</summary>
  <ul>
    <li>rocmPackages.hipblas (rocmPackages_6.hipblas)</li>
    <li>rocmPackages.hipsolver (rocmPackages_6.hipsolver)</li>
    <li>rocmPackages.miopen (rocmPackages.miopen-hip ,rocmPackages_6.miopen ,rocmPackages_6.miopen-hip)</li>
    <li>rocmPackages.rocalution (rocmPackages_6.rocalution)</li>
    <li>rocmPackages.rocblas (rocmPackages_6.rocblas)</li>
    <li>rocmPackages.rocsolver (rocmPackages_6.rocsolver)</li>
  </ul>
</details>
<details>
  <summary>34 packages built:</summary>
  <ul>
    <li>blender-hip</li>
    <li>rocmPackages.clr (rocmPackages_6.clr)</li>
    <li>rocmPackages.clr.icd (rocmPackages_6.clr.icd)</li>
    <li>rocmPackages.composable_kernel (rocmPackages_6.composable_kernel)</li>
    <li>rocmPackages.hipcub (rocmPackages_6.hipcub)</li>
    <li>rocmPackages.hipfft (rocmPackages_6.hipfft)</li>
    <li>rocmPackages.hiprand (rocmPackages.rocrand ,rocmPackages_6.hiprand ,rocmPackages_6.rocrand)</li>
    <li>rocmPackages.hipsparse (rocmPackages_6.hipsparse)</li>
    <li>rocmPackages.llvm.mlir (rocmPackages_6.llvm.mlir)</li>
    <li>rocmPackages.llvm.openmp (rocmPackages_6.llvm.openmp)</li>
    <li>rocmPackages.llvm.openmp.doc (rocmPackages_6.llvm.openmp.doc)</li>
    <li>rocmPackages.llvm.openmp.info (rocmPackages_6.llvm.openmp.info)</li>
    <li>rocmPackages.llvm.openmp.man (rocmPackages_6.llvm.openmp.man)</li>
    <li>rocmPackages.rccl (rocmPackages_6.rccl)</li>
    <li>rocmPackages.rocdbgapi (rocmPackages_6.rocdbgapi)</li>
    <li>rocmPackages.rocdbgapi.doc (rocmPackages_6.rocdbgapi.doc)</li>
    <li>rocmPackages.rocfft (rocmPackages_6.rocfft)</li>
    <li>rocmPackages.rocm-runtime (rocmPackages_6.rocm-runtime)</li>
    <li>rocmPackages.rocminfo (rocmPackages_6.rocminfo)</li>
    <li>rocmPackages.rocmlir (rocmPackages_6.rocmlir)</li>
    <li>rocmPackages.rocmlir-rock (rocmPackages_6.rocmlir-rock)</li>
    <li>rocmPackages.rocmlir.external (rocmPackages_6.rocmlir.external)</li>
    <li>rocmPackages.rocprim (rocmPackages_6.rocprim)</li>
    <li>rocmPackages.rocprofiler (rocmPackages_6.rocprofiler)</li>
    <li>rocmPackages.rocr-debug-agent (rocmPackages_6.rocr-debug-agent)</li>
    <li>rocmPackages.rocsparse (rocmPackages_6.rocsparse)</li>
    <li>rocmPackages.rocthrust (rocmPackages_6.rocthrust)</li>
    <li>rocmPackages.roctracer (rocmPackages_6.roctracer)</li>
    <li>rocmPackages.rocwmma (rocmPackages_6.rocwmma)</li>
    <li>rocmPackages.rpp (rocmPackages.rpp-hip ,rocmPackages_6.rpp ,rocmPackages_6.rpp-hip)</li>
    <li>rocmPackages.rpp-cpu (rocmPackages_6.rpp-cpu)</li>
    <li>rocmPackages.rpp-opencl (rocmPackages_6.rpp-opencl)</li>
    <li>rocmPackages.tensile (rocmPackages_6.tensile)</li>
    <li>rocmPackages.tensile.dist (rocmPackages_6.tensile.dist)</li>
  </ul>
</details>

Here is the build log:

```
@nix { "action": "setPhase", "phase": "unpackPhase" }
Running phase: unpackPhase
unpacking source archive /nix/store/ihrai11dd7zmmvj6j9g1kyj4337f85hh-source
source root is source
@nix { "action": "setPhase", "phase": "patchPhase" }
Running phase: patchPhase
applying patch /nix/store/04x9sfq554jgzhafdkpyznpk0azglwp0-Extend-rocBLAS-HIP-ISA-compatibility.patch
patching file library/src/handle.cpp
patching file library/src/rocblas_auxiliary.cpp
patching file library/src/tensile_host.cpp
ln: failed to create symbolic link 'build/Tensile/library/TensileLibrary.dat': File exists
/nix/store/hylcibqj87y8flsj51d2dpc7fiy6g1gx-stdenv-linux/setup: line 131: pop_var_context: head of shell_variables not a function context
```

Did you manage to run `nixpkgs-review` or are you having trouble with it?
My own experience is that ROCm is quite a pain to build with its huge hardware requirements.

GZGavinZhao commented:
@mschwaig Thank you for the run! I'm still fixing `rocblas`. I realized I could set `gpuTargets` to only my GPU and I had been doing that to reduce the build time/memory requirement.

I had a fix for rocBLAS but it was really messy. I think I figured out a clean solution that requires just one short patch and no more `postPatch`. Currently building locally to verify.

GZGavinZhao commented:
Quick question: I want to test rocBLAS through llama.cpp. How do I build a `llama-cpp` that has `rocmSupport` set to `true`?

mschwaig commented:
> Quick question: I want to test rocBLAS through llama.cpp. How do I build a `llama-cpp` that has `rocmSupport` set to `true`?

Should be `llama-cpp.override { rocmSupport = true; }`

GZGavinZhao committed:
rocmPackages: cleanup and remove double building for rocblas

Signed-off-by: Gavin Zhao <git@gzgz.dev>

GZGavinZhao committed:
rocmPackages: don't build arches in parallel in rccl

Signed-off-by: Gavin Zhao <git@gzgz.dev>

GZGavinZhao commented:
Hi @mschwaig would you mind running `nixpkgs-review` on your end? `rocblas` now compiles fine, but for some reason even 64GB of RAM is not enough to run `nixpkgs-review` and I still frequently get out-of-disk-space errors :(

mschwaig commented:
> Hi @mschwaig would you mind running `nixpkgs-review` on your end? `rocblas` now compiles fine, but for some reason even 64GB of RAM is not enough to run `nixpkgs-review` and I still frequently get out-of-disk-space errors :(

Great to hear that it compiles.

I'm running `nixpkgs-review` now and I will post the results when it completes.

Tungsten842 commented:
I was reading the llvm [documentation](https://llvm.org/docs/AMDGPUUsage.html#id115) for the upcoming generic targets and it doesn't look like that this patch will always work, for example:
![image](https://github.com/NixOS/nixpkgs/assets/24614168/75f738a7-560d-496b-b809-c47a405b95c1)
It looks like that some targets do not support some instructions, so they are not really compatible. So for example for rocblas with this patch if you are using a gfx1012 by default you will be using gfx1010 code, that includes some instructions that are not supported by gfx1012. It looks like this might cause some problems.

mschwaig commented:
Result of `nixpkgs-review pr 298388` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>14 packages marked as broken and skipped:</summary>
  <ul>
    <li>rocmPackages.llvm.flang</li>
    <li>rocmPackages.llvm.flang.doc</li>
    <li>rocmPackages.llvm.flang.info</li>
    <li>rocmPackages.llvm.flang.man</li>
    <li>rocmPackages.migraphx</li>
    <li>rocmPackages.rdc</li>
    <li>rocmPackages.rdc.doc</li>
    <li>rocmPackages_6.llvm.flang</li>
    <li>rocmPackages_6.llvm.flang.doc</li>
    <li>rocmPackages_6.llvm.flang.info</li>
    <li>rocmPackages_6.llvm.flang.man</li>
    <li>rocmPackages_6.migraphx</li>
    <li>rocmPackages_6.rdc</li>
    <li>rocmPackages_6.rdc.doc</li>
  </ul>
</details>
<details>
  <summary>5 packages failed to build:</summary>
  <ul>
    <li>rocmPackages.hipblas (rocmPackages_6.hipblas)</li>
    <li>rocmPackages.hipsolver (rocmPackages_6.hipsolver)</li>
    <li>rocmPackages.miopen (rocmPackages.miopen-hip ,rocmPackages_6.miopen ,rocmPackages_6.miopen-hip)</li>
    <li>rocmPackages.rocalution (rocmPackages_6.rocalution)</li>
    <li>rocmPackages.rocsolver (rocmPackages_6.rocsolver)</li>
  </ul>
</details>
<details>
  <summary>35 packages built:</summary>
  <ul>
    <li>blender-hip</li>
    <li>rocmPackages.clr (rocmPackages_6.clr)</li>
    <li>rocmPackages.clr.icd (rocmPackages_6.clr.icd)</li>
    <li>rocmPackages.composable_kernel (rocmPackages_6.composable_kernel)</li>
    <li>rocmPackages.hipcub (rocmPackages_6.hipcub)</li>
    <li>rocmPackages.hipfft (rocmPackages_6.hipfft)</li>
    <li>rocmPackages.hiprand (rocmPackages.rocrand ,rocmPackages_6.hiprand ,rocmPackages_6.rocrand)</li>
    <li>rocmPackages.hipsparse (rocmPackages_6.hipsparse)</li>
    <li>rocmPackages.llvm.mlir (rocmPackages_6.llvm.mlir)</li>
    <li>rocmPackages.llvm.openmp (rocmPackages_6.llvm.openmp)</li>
    <li>rocmPackages.llvm.openmp.doc (rocmPackages_6.llvm.openmp.doc)</li>
    <li>rocmPackages.llvm.openmp.info (rocmPackages_6.llvm.openmp.info)</li>
    <li>rocmPackages.llvm.openmp.man (rocmPackages_6.llvm.openmp.man)</li>
    <li>rocmPackages.rccl (rocmPackages_6.rccl)</li>
    <li>rocmPackages.rocblas (rocmPackages_6.rocblas)</li>
    <li>rocmPackages.rocdbgapi (rocmPackages_6.rocdbgapi)</li>
    <li>rocmPackages.rocdbgapi.doc (rocmPackages_6.rocdbgapi.doc)</li>
    <li>rocmPackages.rocfft (rocmPackages_6.rocfft)</li>
    <li>rocmPackages.rocm-runtime (rocmPackages_6.rocm-runtime)</li>
    <li>rocmPackages.rocminfo (rocmPackages_6.rocminfo)</li>
    <li>rocmPackages.rocmlir (rocmPackages_6.rocmlir)</li>
    <li>rocmPackages.rocmlir-rock (rocmPackages_6.rocmlir-rock)</li>
    <li>rocmPackages.rocmlir.external (rocmPackages_6.rocmlir.external)</li>
    <li>rocmPackages.rocprim (rocmPackages_6.rocprim)</li>
    <li>rocmPackages.rocprofiler (rocmPackages_6.rocprofiler)</li>
    <li>rocmPackages.rocr-debug-agent (rocmPackages_6.rocr-debug-agent)</li>
    <li>rocmPackages.rocsparse (rocmPackages_6.rocsparse)</li>
    <li>rocmPackages.rocthrust (rocmPackages_6.rocthrust)</li>
    <li>rocmPackages.roctracer (rocmPackages_6.roctracer)</li>
    <li>rocmPackages.rocwmma (rocmPackages_6.rocwmma)</li>
    <li>rocmPackages.rpp (rocmPackages.rpp-hip ,rocmPackages_6.rpp ,rocmPackages_6.rpp-hip)</li>
    <li>rocmPackages.rpp-cpu (rocmPackages_6.rpp-cpu)</li>
    <li>rocmPackages.rpp-opencl (rocmPackages_6.rpp-opencl)</li>
    <li>rocmPackages.tensile (rocmPackages_6.tensile)</li>
    <li>rocmPackages.tensile.dist (rocmPackages_6.tensile.dist)</li>
  </ul>
</details>

These are the logs of the failing packages:
[rocmPackages.rocsolver.log](https://github.com/NixOS/nixpkgs/files/14804082/rocmPackages.rocsolver.log)
[rocmPackages.rocalution.log](https://github.com/NixOS/nixpkgs/files/14804083/rocmPackages.rocalution.log)
[rocmPackages.miopen.log](https://github.com/NixOS/nixpkgs/files/14804084/rocmPackages.miopen.log)

The logs for `hipblas` and `hipsolver` were empty since their dependencies failed to build.

GZGavinZhao committed:
rocmPackages: install headers correctly in rocblas

Signed-off-by: Gavin Zhao <git@gzgz.dev>

GZGavinZhao commented:
Hi @Tungsten842, I believe it's actually the reverse. The document is saying that _because_ we're compiling against `gfx1010`, when you're running an executable compiled against `gfx1010` on `gfx1012` hardware, the instruction `v_dot4_i32_i8` would not be available for you to use (it's there, but your executable would never use this instruction) since the compiler has to assume you may be running on `gfx1010`. For example, you can see from the [`gfx1010` instruction docs](https://llvm.org/docs/AMDGPU/AMDGPUAsmGFX10.html) that `v_dot4_i32_i8` is not present, but this instruction is present in the [`gfx1011`/`gfx1012` instruction docs](https://llvm.org/docs/AMDGPU/AMDGPUAsmGFX1011.html).

GZGavinZhao commented:
@mschwaig I believe I have fixed the issue (the header file `rocblas.h` is now being installed properly). Would you mind running `nixpkgs-review` when you get a chance? Thanks!

Tungsten842 commented:
> Hi @Tungsten842, I believe it's actually the reverse. The document is saying that _because_ we're compiling against `gfx1010`, when you're running an executable compiled against `gfx1010` on `gfx1012` hardware, the instruction `v_dot4_i32_i8` would not be available for you to use (the instruction would work properly if you try to call it, but your executable would never use this instruction) since the compiler has to assume you may be running on `gfx1010`. For example, you can see from the [`gfx1010` instruction docs](https://llvm.org/docs/AMDGPU/AMDGPUAsmGFX10.html) that `v_dot4_i32_i8` is not present, but this instruction is present in the [`gfx1011`/`gfx1012` instruction docs](https://llvm.org/docs/AMDGPU/AMDGPUAsmGFX1011.html).
> 
> Therefore, `gfx1010` is indeed the greatest common denominator for RDNA1 GPUs and all RDNA1 GPUs can use `gfx1010` instructions as the basis ;)

Makes sense. But there is still one problem, it looks like that gfx900 supports some v_mad_mix instructions. But (gfx906, gfx904...) do not.
https://llvm.org/docs/AMDGPU/AMDGPUAsmGFX900.html
https://llvm.org/docs/AMDGPU/AMDGPUAsmGFX904.html
https://llvm.org/docs/AMDGPU/AMDGPUAsmGFX906.html
Also gfx1100 seems to have more VGPRs than other targets, not sure about the impact of this: https://reviews.llvm.org/D134522.

GZGavinZhao commented:
Finally, `llama-cpp` has been built. I am able to run Gemma 2B on `gfx1032` without setup or environment variables. It just works™

mschwaig commented:
> @mschwaig I believe I have fixed the issue (the header file `rocblas.h` is now being installed properly). Would you mind running `nixpkgs-review` when you get a chance? Thanks!

Sure, no problem. I am running it now.

Congrats on getting `llama.cpp` to work. I think my suggestion of using `nix eval` was not that good, because you only get very minimal output that way.

I read a comment on an article recently where someone from AMD left a comment about supporting a wider range of hardware: https://www.phoronix.com/forums/forum/linux-graphics-x-org-drivers/open-source-amd-linux/1446020-amd-preparing-rocm-6-1-for-release-with-new-features?p=1446093#post1446093

That comment makes me think that if we have questions about what ISAs to target, we could also consider contacting someone at AMD and asking.

mschwaig commented:
Result of `nixpkgs-review pr 298388` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>14 packages marked as broken and skipped:</summary>
  <ul>
    <li>rocmPackages.llvm.flang</li>
    <li>rocmPackages.llvm.flang.doc</li>
    <li>rocmPackages.llvm.flang.info</li>
    <li>rocmPackages.llvm.flang.man</li>
    <li>rocmPackages.migraphx</li>
    <li>rocmPackages.rdc</li>
    <li>rocmPackages.rdc.doc</li>
    <li>rocmPackages_6.llvm.flang</li>
    <li>rocmPackages_6.llvm.flang.doc</li>
    <li>rocmPackages_6.llvm.flang.info</li>
    <li>rocmPackages_6.llvm.flang.man</li>
    <li>rocmPackages_6.migraphx</li>
    <li>rocmPackages_6.rdc</li>
    <li>rocmPackages_6.rdc.doc</li>
  </ul>
</details>
<details>
  <summary>40 packages built:</summary>
  <ul>
    <li>blender-hip</li>
    <li>rocmPackages.clr (rocmPackages_6.clr)</li>
    <li>rocmPackages.clr.icd (rocmPackages_6.clr.icd)</li>
    <li>rocmPackages.composable_kernel (rocmPackages_6.composable_kernel)</li>
    <li>rocmPackages.hipblas (rocmPackages_6.hipblas)</li>
    <li>rocmPackages.hipcub (rocmPackages_6.hipcub)</li>
    <li>rocmPackages.hipfft (rocmPackages_6.hipfft)</li>
    <li>rocmPackages.hiprand (rocmPackages.rocrand ,rocmPackages_6.hiprand ,rocmPackages_6.rocrand)</li>
    <li>rocmPackages.hipsolver (rocmPackages_6.hipsolver)</li>
    <li>rocmPackages.hipsparse (rocmPackages_6.hipsparse)</li>
    <li>rocmPackages.llvm.mlir (rocmPackages_6.llvm.mlir)</li>
    <li>rocmPackages.llvm.openmp (rocmPackages_6.llvm.openmp)</li>
    <li>rocmPackages.llvm.openmp.doc (rocmPackages_6.llvm.openmp.doc)</li>
    <li>rocmPackages.llvm.openmp.info (rocmPackages_6.llvm.openmp.info)</li>
    <li>rocmPackages.llvm.openmp.man (rocmPackages_6.llvm.openmp.man)</li>
    <li>rocmPackages.miopen (rocmPackages.miopen-hip ,rocmPackages_6.miopen ,rocmPackages_6.miopen-hip)</li>
    <li>rocmPackages.rccl (rocmPackages_6.rccl)</li>
    <li>rocmPackages.rocalution (rocmPackages_6.rocalution)</li>
    <li>rocmPackages.rocblas (rocmPackages_6.rocblas)</li>
    <li>rocmPackages.rocdbgapi (rocmPackages_6.rocdbgapi)</li>
    <li>rocmPackages.rocdbgapi.doc (rocmPackages_6.rocdbgapi.doc)</li>
    <li>rocmPackages.rocfft (rocmPackages_6.rocfft)</li>
    <li>rocmPackages.rocm-runtime (rocmPackages_6.rocm-runtime)</li>
    <li>rocmPackages.rocminfo (rocmPackages_6.rocminfo)</li>
    <li>rocmPackages.rocmlir (rocmPackages_6.rocmlir)</li>
    <li>rocmPackages.rocmlir-rock (rocmPackages_6.rocmlir-rock)</li>
    <li>rocmPackages.rocmlir.external (rocmPackages_6.rocmlir.external)</li>
    <li>rocmPackages.rocprim (rocmPackages_6.rocprim)</li>
    <li>rocmPackages.rocprofiler (rocmPackages_6.rocprofiler)</li>
    <li>rocmPackages.rocr-debug-agent (rocmPackages_6.rocr-debug-agent)</li>
    <li>rocmPackages.rocsolver (rocmPackages_6.rocsolver)</li>
    <li>rocmPackages.rocsparse (rocmPackages_6.rocsparse)</li>
    <li>rocmPackages.rocthrust (rocmPackages_6.rocthrust)</li>
    <li>rocmPackages.roctracer (rocmPackages_6.roctracer)</li>
    <li>rocmPackages.rocwmma (rocmPackages_6.rocwmma)</li>
    <li>rocmPackages.rpp (rocmPackages.rpp-hip ,rocmPackages_6.rpp ,rocmPackages_6.rpp-hip)</li>
    <li>rocmPackages.rpp-cpu (rocmPackages_6.rpp-cpu)</li>
    <li>rocmPackages.rpp-opencl (rocmPackages_6.rpp-opencl)</li>
    <li>rocmPackages.tensile (rocmPackages_6.tensile)</li>
    <li>rocmPackages.tensile.dist (rocmPackages_6.tensile.dist)</li>
  </ul>
</details>

GZGavinZhao commented:
Thank you @mschwaig. I think we're good to go!

mschwaig reviewed (changes requested):
This PR is really interesting, thanks.

### Review result summary

I have reviewed your changes as well as the referenced patches, so I have no concerns about testing this on my own machine.

My comments are mostly about
* smaller issues in the code,
* a need for explanation on some specifics in the code and
* questions about the origin of some of the patches.

This looks like quite a big change to me, which should get as much testing as possible before it gets merged. I will start running these changes on my system as soon as I can.

### My view on where these changes are taking us

ROCm only runs on GPUs they do verification on, and personally I think it makes sense for nixpkgs to merge something like this and distribute a version of ROCm which supports a wider range of devices.

It seems like this means we have to carry the changes from upstream that this introduces forward and adapt them for new generations of GPUs, which is a maintenance burden.
With this kind of thing it would be great if there was as little duplicated effort between distributions as possible. It should be doable for people updating ROCm in the future to update this.

Overall I think it would be great if we end up merging this.

### Questions

I have a few more questions that I did not think fit the review comment format:

1. Do you want to upstream this?
2. Is something like this is likely to land upstream?
3. Who else is carrying or planning to carry these patches? Solus?
4. Do these changes cover all of ROCm?
5. Is `composable_kernel` affected by these changes (since that project also seems to build a lot of things per ISA)? How?

GZGavinZhao commented:
@mschwaig Thank you for your review! I'll respond to the review comments either tomorrow or Monday. I'll respond to your questions at the end of your comment first since I'm asking another person to join this conversation.

> 1. Do you want to upstream this?
> 2. Is something like this is likely to land upstream?

I'll answer these two questions together: No to both. To my knowledge, AMD has already presented and upstreamed (to LLVM) [Generic Architectures](https://github.com/llvm/llvm-project/pull/76955) as the once-and-for-all and robust solution to GPU compatibility issues. Therefore, they will not be accepting workaround patches like the ones presented in this PR.

Hi @cgmb. Sorry to ping you. If you could verify the info above, that would be appreciated. Thank you!

> 3. Who else is carrying or planning to carry these patches? Solus?

Solus definitely yes, as long as I'm the ROCm maintainer for Solus. Probably Debian as well depending on @cgmb. Ideally I would like to convince Arch and Fedora to also get on to the same boat. :smiling_imp: 

> 4. Do these changes cover all of ROCm?

To the best of my knowledge, yes. The changes in `clr` and `rocm-runtime` is enough to cover 90% of the use cases, because it directly targets the HIP runtime. The only cases where we still need to patch manually (rocBLAS, Tensile, and MIOpen) is caused by these libraries querying the GPU arch **on their own** to choose what GPU code to execute, instead of relying on the HIP runtime to decide.

`llama.cpp`, stable diffusion, and any application/library that relies on ROCm that I have run, can run with these set of patches without issues. My hardware is `gfx1032` (masked as `gfx1030`) and `gfx90c` (masked as `gfx900`).

> 5. Is `composable_kernel` affected by these changes (since that project also seems to build a lot of things per ISA)? How?

No. As explained in the response in 1. and 2., the changes in the HIP runtime covers most cases, which includes `composable_kernel`. I have read through the codebase of `composable_kernel` and concluded that `composable_kernel` relies solely on the HIP runtime to decide what GPU code to execute, so patches to the HIP runtime is enough. I have verified this by passing the `composable_kernel` test suite on my GPU with this set of patches.

cgmb commented on pkgs/development/rocm-modules/6/clr/default.nix:
```diff
@@ -100,6 +100,16 @@ in stdenv.mkDerivation (finalAttrs: {
+      hash = "sha256-kbEeJsQgAxbNfBCPB2Jny4z526UAnIxHNQLGsD2iFvg=";
+    })
+    (fetchpatch {
+      name = "improve-rocclr-isa-compatibility-check.patch";
```
> Where is this from?

I wrote the patch for Debian.
https://salsa.debian.org/rocm-team/rocm-hipamd/-/blob/debian/5.7.1-3/debian/patches/0025-improve-rocclr-isa-compatibility-check.patch

> Are we allowed to include this?

Yes.

GZGavinZhao replied:
Thank you @cgmb! @mschwaig to prevent confusion, I've now switched the patches' url to Debian whenever possible. The only ones that are not switched are those that don't apply cleanly and needs manual fixes/rebase.

mschwaig replied:
This looks good.

For the ´rocclr isa compatibility check´, are you aware that the new patch is slightly different?

Just asking to make sure this is intended.

```diff
1d0
< From 2783c57b0f225ad8bc553e2d244837d57d8375bc Mon Sep 17 00:00:00 2001
4c3
< Subject: [PATCH] improve rocclr isa compatibility check
---
> Subject: improve rocclr isa compatibility check
17,18d15
< 
< Signed-off-by: Gavin Zhao <git@gzgz.dev>
20,21c17,18
<  rocclr/device/device.cpp | 39 ++++++++++++++++++++++++++++++++++++---
<  1 file changed, 36 insertions(+), 3 deletions(-)
---
>  rocclr/device/device.cpp | 45 ++++++++++++++++++++++++++++++++++++++++++---
>  1 file changed, 42 insertions(+), 3 deletions(-)
23,24d19
< diff --git a/rocclr/device/device.cpp b/rocclr/device/device.cpp
< index 0249f31d6..739bb027c 100644
27c22
< @@ -234,10 +234,43 @@ std::string Isa::isaName() const {
---
> @@ -232,10 +232,49 @@
49,51c44,47
< +        const std::array<uint32_t, 4> equivalent_gfx90x = { 0, 2, 9, 12 };
< +        if (Contains(equivalent_gfx90x, codeObjectIsa.versionStepping()) &&
< +            Contains(equivalent_gfx90x, agentIsa.versionStepping())) {
---
> +      const std::array<uint32_t, 4> gfx900_equivalent = { 0, 2, 9, 12 };
> +      const std::array<uint32_t, 5> gfx900_superset = { 0, 2, 6, 9, 12 };
> +      if (Contains(gfx900_equivalent, codeObjectIsa.versionStepping()) &&
> +          Contains(gfx900_superset, agentIsa.versionStepping())) {
55a52,55
> +        const std::array<uint32_t, 1> gfx1010_equivalent = { 0 };
> +        const std::array<uint32_t, 4> gfx1010_superset = { 0, 1, 2, 3 };
> +        if (Contains(gfx1010_equivalent, codeObjectIsa.versionStepping()) &&
> +            Contains(gfx1010_superset, agentIsa.versionStepping())) {
56a57
> +        }
```

GZGavinZhao replied:
Yes, I'm aware. The new patch is just being more cautious on compatibility by checking both the code object ISA and the host ISA. The difference is intended.

cgmb commented:
> > 1. Do you want to upstream this?
> > 2. Is something like this is likely to land upstream?
> 
> I'll answer these two questions together: No to both. To my knowledge, AMD has already presented and upstreamed (to LLVM) [Generic Targets](https://github.com/llvm/llvm-project/pull/76955) as the once-and-for-all and robust solution to GPU compatibility issues. Therefore, they will not be accepting workaround patches like the ones presented in this PR.
> 
> Hi @cgmb. Sorry to ping you. If you could verify the info above, that would be appreciated. Thank you!

That is correct. I proposed these patches upstream and they were explicitly rejected in favour of pushing the compatibility logic into the compiler. I was consulted during the development of the LLVM generic targets and I agree that they are the better solution. However, it will be some time before generic targets are supported by the HIP runtime and the ROCm libraries.

I neither endorse nor condemn using these patches on NixOS. I supported merging them into Debian Unstable because the upstream solution would not be available in time for Ubuntu 24.04. I will drop them from Debian the next time the libamdhip64 soname is changed, if the upstream solution is available at that time.

Note that I have updated the patches slightly after reviewing the LLVM generic ISAs. The gfx1012 ISA is a superset of gfx1010, not equivalent to gfx1010.

cgmb commented:
> > 7\. Do these changes cover all of ROCm?
> 
> To the best of my knowledge, yes.

There are some dynamic checks in rocprim (at least in 5.5 and later). I added [a patch](https://salsa.debian.org/rocm-team/rocprim/-/blob/debian/5.7.1-2/debian/patches/arch-conversion-macro.patch) to allow users to opt-in at compile time to treating gfx1031 through gfx1036 as gfx1030; and gfx902, gfx909, and gfx90c as gfx900.

> ROCm only runs on GPUs they do verification on, and personally I think it makes sense for nixpkgs to merge something like this and distribute a version of ROCm which supports a wider range of devices.

FWIW, I've acquired a wide range of test hardware for Debian. It's not all available on our CI yet, but we run quite a lot of tests on GPUs that are not officially supported upstream. Debian ROCm CI: https://ci.rocm.debian.net/ Test CI (for upcoming systems): https://ci-test.rocm.debian.net/

mschwaig commented:
> No, `composable_kernel` doesn't need additional changes to get the same compatibility improvement. As explained in the response in 4., the changes in the HIP runtime covers most cases, which includes `composable_kernel`. I have read through the codebase of `composable_kernel` and concluded that `composable_kernel` relies solely on the HIP runtime to decide what GPU code to execute, so patches to the HIP runtime is enough. I have verified this by passing the `composable_kernel` test suite on my GPU with this set of patches.

Could we drop some of the default GPU targets from `composable_kernel` and just have a small performance loss on the dropped targets?

A build for all default targets  is really large right now (large enough to cause issues https://github.com/NixOS/nixpkgs/pull/299589), because of
`${composable_kernel}/lib/libdevice_operations.a`, weighing in at 3.7 GB with `DEFAULT_GPU_TARGETS=gfx908;gfx90a;gfx940;gfx941;gfx942;gfx1030;gfx1100;gfx1101;gfx1102`.

When I, for example, just compile for `gpuTargets = ["gfx908" "gfx1030" "gfx1100"]` that goes down to 1.9 GB. The build times are also much lower of course.

cgmb commented:
> Could we drop some of the default GPU targets from `composable_kernel` and just have a small performance loss on the dropped targets?
> 
> A build for all default targets is really large right now (large enough to cause issues #299589), because of `${composable_kernel}/lib/libdevice_operations.a`, weighing in at 3.7 GB with `DEFAULT_GPU_TARGETS=gfx908;gfx90a;gfx940;gfx941;gfx942;gfx1030;gfx1100;gfx1101;gfx1102`.

No, those ISAs all have meaningful differences between them. It might be possible to replace gfx1100, gfx1101 and gfx1102 with gfx11-generic in the future.

If you need to cut down on size, you could consider dropping gfx940 and gfx941. Those ISAs were used for early engineering samples, but all final MI300 hardware uses the gfx942 ISA.

GZGavinZhao commented on pkgs/development/rocm-modules/6/rccl/default.nix:
```diff
@@ -65,7 +65,9 @@ stdenv.mkDerivation (finalAttrs: {
     substituteInPlace CMakeLists.txt \
-      --replace "\''$ \''${hipify-perl_executable}" "${perl}/bin/perl ${hipify}/bin/hipify-perl"
+      --replace "\''$ \''${hipify-perl_executable}" "${perl}/bin/perl ${hipify}/bin/hipify-perl" \
+      --replace-warn "-parallel-jobs=12" "-parallel-jobs=1" \
+      --replace-warn "-parallel-jobs=16" "-parallel-jobs=1"
```
To respect `NIX_BUILD_CORES`. If not changed, the HIP compiler would launch 12 (or 16) threads _for each build job_ to compile GPU architectures in parallel. This cause the actual cores usage to be `NIX_BUILD_CORES`*12 or *16, which would likely cause memory issues. By setting them to 1 (which is the default), we restore the compiler back to the default behavior and therefore adhere to `NIX_BUILD_CORES`, preventing surprising CPU and memory usage when building.

GZGavinZhao commented on pkgs/development/rocm-modules/6/tensile/default.nix:
```diff
@@ -34,6 +35,19 @@ buildPythonPackage rec {
+    })
+    (fetchpatch {
+      name = "Don-t-copy-file-twice-in-copyStaticFiles.patch";
+      url = "https://github.com/GZGavinZhao/Tensile/commit/9e14d5a00a096bddac605910a0e4dfb4c35bb0d5.patch";
```
In the original rocBLAS recipe, Tensile was rewrapped because `# Tensile REALLY wants to write to the nix directory if we include it normally`. More precisely, during setup phase, Tensile copies the C++ headers it provides into the build directory of rocBLAS. Then, during build phase, Tensile copies it again. _However_, for some reason, despite that it is copying the same file, copying would fail with permission error during build phase but not setup phase. I'm very confused by why this happens, but I've logged the copy commands and verified that it is indeed copying the same file. This patch basically let Tensile only copy the headers if source and destination have different contents. So yes, this check is specifically for nixpkgs to allow us not needing to rewrap Tensile when building `rocblas`.

GZGavinZhao replied:
I'd also note that I can't build rocBLAS on the `master` branch because rocBLAS cannot find Tensile during setup phase, so that's another reason to use this patch and not wrap Tensile anymore in rocBLAS.

mschwaig replied:
This makes sense.

I think prevents source from being writable in later build phases on purpose.

GZGavinZhao commented on pkgs/development/rocm-modules/6/rocblas/default.nix:
> Why is it not necessary anymore to call the tensile build separately for each GPU generation and merge the results?

1. It shouldn't have been necessary in the first place.
2. The way of building separately for each architecture breaks when `sepArchitectures` is false (as probably implied by the name). Building architectures together fixed it.

mschwaig replied:
I think the reason for this was that without it `rocblas` is slightly too large for the cache.

```
$ nix build .#rocmPackages.rocblas
$ du -h --apparent-size --max-depth 0 result/
3.1G	result/
```

This is something we could address via https://github.com/NixOS/nixpkgs/pull/305920 or something like https://github.com/NixOS/nixpkgs/pull/299589/commits/e380b5303715016fca18df860d97340345b52aef, but I don't know if we need a solution for the size issue before merging this, since we're not making things worse than they are right now anyways.

I have it on my list to check the sizes of other outputs.

GZGavinZhao replied:
I'd say merge this for now then fix it later, since `rocblas` doesn't even build on master right now. Then either #305920 will fix it or we can figure out how to do splitting.

ulrikstrid replied:
Isn't it broken on master because of cmake? Seems like we don't want to introduce a cache miss just because it's broken on master.

mschwaig replied:
You're right, it's fixed on master now.

If we merge this PR as is `rocblas` will no longer be cached.

I think a better argument might be that not having `composable_kernel` in cache is a much bigger issue than not having `rocblas` in cache, since
1. `rocblas` only takes 20 minutes to build, orders of magnitude less than `composable_kernel`.
2. Whatever solution we come up with to cache `composable_kernel`, we can also use to cache `rocblas`.

Probably that argument is also invalid to some degree, because  `composable_kernel` is not a runtime dependency of `rocblas`, and applications depend on `rocblas`, not `composable_kernel`.
So I think as long as `rocblas` is cached most users do not suffer from `composable_kernel` not being cached.

Is that right?

mschwaig replied:
I am now checking if removing the early-engineering sample only gfx940 and gfx941 targets from just `rocblas` would get us below 3 GB.

mschwaig replied:
Yes, this would get us down to 2.7GB: https://github.com/NixOS/nixpkgs/commit/89ab15f76cbafe1c1166b05bf4e49298879db998

GZGavinZhao committed:
rocmPackages: fix dependencies in rocblas

Signed-off-by: Gavin Zhao <git@gzgz.dev>

GZGavinZhao commented on pkgs/development/rocm-modules/6/clr/default.nix:
```diff
@@ -100,6 +100,16 @@ in stdenv.mkDerivation (finalAttrs: {
     })
+    (fetchpatch {
+      name = "Improve-hipamd-compat-check.patch";
+      url = "https://github.com/GZGavinZhao/clr/commit/f52172a0767f88bf386dc615a3354156d023bdb8.patch";
```
I don't know, it's @cgmb's code. I think he was just trying to short-circuit. It works so I didn't think much on it :sweat_smile:

GZGavinZhao committed:
rocmPackages: fetch patches from Debian if possible

Signed-off-by: Gavin Zhao <git@gzgz.dev>

GZGavinZhao commented:
> There are some dynamic checks in rocprim (at least in 5.5 and later). I added [a patch](https://salsa.debian.org/rocm-team/rocprim/-/blob/debian/5.7.1-2/debian/patches/arch-conversion-macro.patch) to allow users to opt-in at compile time to treating gfx1031 through gfx1036 as gfx1030; and gfx902, gfx909, and gfx90c as gfx900.

Patch has been included. Thanks for the catch!

GZGavinZhao committed:
rocmPackages: add isa conversion patch in rocprim

Signed-off-by: Gavin Zhao <git@gzgz.dev>

GZGavinZhao committed:
rocmPackages: no need to pass rocblas to itself

Signed-off-by: Gavin Zhao <git@gzgz.dev>

mschwaig commented:
Result of `nixpkgs-review pr 298388` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>20 packages marked as broken and skipped:</summary>
  <ul>
    <li>rocmPackages.llvm.flang</li>
    <li>rocmPackages.llvm.flang.doc</li>
    <li>rocmPackages.llvm.flang.info</li>
    <li>rocmPackages.llvm.flang.man</li>
    <li>rocmPackages.migraphx</li>
    <li>rocmPackages.mivisionx</li>
    <li>rocmPackages.mivisionx-cpu</li>
    <li>rocmPackages.mivisionx-hip</li>
    <li>rocmPackages.rdc</li>
    <li>rocmPackages.rdc.doc</li>
    <li>rocmPackages_6.llvm.flang</li>
    <li>rocmPackages_6.llvm.flang.doc</li>
    <li>rocmPackages_6.llvm.flang.info</li>
    <li>rocmPackages_6.llvm.flang.man</li>
    <li>rocmPackages_6.migraphx</li>
    <li>rocmPackages_6.mivisionx</li>
    <li>rocmPackages_6.mivisionx-cpu</li>
    <li>rocmPackages_6.mivisionx-hip</li>
    <li>rocmPackages_6.rdc</li>
    <li>rocmPackages_6.rdc.doc</li>
  </ul>
</details>
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>rocmPackages.rocprofiler (rocmPackages_6.rocprofiler)</li>
  </ul>
</details>
<details>
  <summary>39 packages built:</summary>
  <ul>
    <li>blender-hip</li>
    <li>rocmPackages.clr (rocmPackages_6.clr)</li>
    <li>rocmPackages.clr.icd (rocmPackages_6.clr.icd)</li>
    <li>rocmPackages.composable_kernel (rocmPackages_6.composable_kernel)</li>
    <li>rocmPackages.hipblas (rocmPackages_6.hipblas)</li>
    <li>rocmPackages.hipcub (rocmPackages_6.hipcub)</li>
    <li>rocmPackages.hipfft (rocmPackages_6.hipfft)</li>
    <li>rocmPackages.hiprand (rocmPackages.rocrand ,rocmPackages_6.hiprand ,rocmPackages_6.rocrand)</li>
    <li>rocmPackages.hipsolver (rocmPackages_6.hipsolver)</li>
    <li>rocmPackages.hipsparse (rocmPackages_6.hipsparse)</li>
    <li>rocmPackages.llvm.mlir (rocmPackages_6.llvm.mlir)</li>
    <li>rocmPackages.llvm.openmp (rocmPackages_6.llvm.openmp)</li>
    <li>rocmPackages.llvm.openmp.doc (rocmPackages_6.llvm.openmp.doc)</li>
    <li>rocmPackages.llvm.openmp.info (rocmPackages_6.llvm.openmp.info)</li>
    <li>rocmPackages.llvm.openmp.man (rocmPackages_6.llvm.openmp.man)</li>
    <li>rocmPackages.miopen (rocmPackages.miopen-hip ,rocmPackages_6.miopen ,rocmPackages_6.miopen-hip)</li>
    <li>rocmPackages.rccl (rocmPackages_6.rccl)</li>
    <li>rocmPackages.rocalution (rocmPackages_6.rocalution)</li>
    <li>rocmPackages.rocblas (rocmPackages_6.rocblas)</li>
    <li>rocmPackages.rocdbgapi (rocmPackages_6.rocdbgapi)</li>
    <li>rocmPackages.rocdbgapi.doc (rocmPackages_6.rocdbgapi.doc)</li>
    <li>rocmPackages.rocfft (rocmPackages_6.rocfft)</li>
    <li>rocmPackages.rocm-runtime (rocmPackages_6.rocm-runtime)</li>
    <li>rocmPackages.rocminfo (rocmPackages_6.rocminfo)</li>
    <li>rocmPackages.rocmlir (rocmPackages_6.rocmlir)</li>
    <li>rocmPackages.rocmlir-rock (rocmPackages_6.rocmlir-rock)</li>
    <li>rocmPackages.rocmlir.external (rocmPackages_6.rocmlir.external)</li>
    <li>rocmPackages.rocprim (rocmPackages_6.rocprim)</li>
    <li>rocmPackages.rocr-debug-agent (rocmPackages_6.rocr-debug-agent)</li>
    <li>rocmPackages.rocsolver (rocmPackages_6.rocsolver)</li>
    <li>rocmPackages.rocsparse (rocmPackages_6.rocsparse)</li>
    <li>rocmPackages.rocthrust (rocmPackages_6.rocthrust)</li>
    <li>rocmPackages.roctracer (rocmPackages_6.roctracer)</li>
    <li>rocmPackages.rocwmma (rocmPackages_6.rocwmma)</li>
    <li>rocmPackages.rpp (rocmPackages.rpp-hip ,rocmPackages_6.rpp ,rocmPackages_6.rpp-hip)</li>
    <li>rocmPackages.rpp-cpu (rocmPackages_6.rpp-cpu)</li>
    <li>rocmPackages.rpp-opencl (rocmPackages_6.rpp-opencl)</li>
    <li>rocmPackages.tensile (rocmPackages_6.tensile)</li>
    <li>rocmPackages.tensile.dist (rocmPackages_6.tensile.dist)</li>
  </ul>
</details>

I think `rocprofiler` just fails on my machine, so the `nixpkgs-review` this is looking good now.
Due to https://github.com/NixOS/nixpkgs/pull/298979#issuecomment-2067629025  building this PR against master fails right now.
I will finish my review of this in the next few days, I wish I would have gotten to it sooner.
Last time i started reviewing this I realized that I had broken ROCm with that patch to add compression. 🤦

I already successfully tested this by with my old RX 6600. 😊

mschwaig reviewed (changes requested):
This is looking good!

I have tested this branch with an RX 6600 and a RX 7600 XT and ROCm attribute of the upstream llama.cpp flake.

There are effectively two comment threads with open questions left.
I think we can probably resolve those without any code changes.

mschwaig reviewed (approved):
I think we can merge this now.

I agree with this comment: https://github.com/NixOS/nixpkgs/pull/298388/files#r1576052202 and there are no other open issues at all.

EDIT: LGTM.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/3826

ulrikstrid reviewed (commented):
I just have a question about the output size that I added to the thread. Other than that the changes seems fine.

GZGavinZhao committed:
rocmPackages.rocblas: do not build for gfx1012, gfx940, and gfx 941 GPU targets

Signed-off-by: Gavin Zhao <git@gzgz.dev>

GZGavinZhao commented:
@mschwaig Updated according to https://github.com/NixOS/nixpkgs/commit/89ab15f76cbafe1c1166b05bf4e49298879db998

cgmb commented:
> @mschwaig Updated according to [89ab15f](https://github.com/NixOS/nixpkgs/commit/89ab15f76cbafe1c1166b05bf4e49298879db998)

You've already patched rocBLAS to run gfx1010 kernels on gfx1012 hardware, so you can further reduce the size by dropping gfx1012 from that list.

mschwaig commented:
> @mschwaig Updated according to [89ab15f](https://github.com/NixOS/nixpkgs/commit/89ab15f76cbafe1c1166b05bf4e49298879db998)

Great! I already ran `nixpkgs-review` again with that change.

Result of `nixpkgs-review` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>10 packages marked as broken and skipped:</summary>
  <ul>
    <li>rocmPackages.llvm.flang</li>
    <li>rocmPackages.llvm.flang.doc</li>
    <li>rocmPackages.llvm.flang.info</li>
    <li>rocmPackages.llvm.flang.man</li>
    <li>rocmPackages.migraphx</li>
    <li>rocmPackages.mivisionx</li>
    <li>rocmPackages.mivisionx-cpu</li>
    <li>rocmPackages.mivisionx-hip</li>
    <li>rocmPackages.rdc</li>
    <li>rocmPackages.rdc.doc</li>
  </ul>
</details>
<details>
  <summary>42 packages built:</summary>
  <ul>
    <li>blender-hip</li>
    <li>rocmPackages.clr</li>
    <li>rocmPackages.clr.icd</li>
    <li>rocmPackages.composable_kernel</li>
    <li>rocmPackages.hipblas</li>
    <li>rocmPackages.hipcub</li>
    <li>rocmPackages.hipfft</li>
    <li>rocmPackages.hiprand</li>
    <li>rocmPackages.hipsolver</li>
    <li>rocmPackages.hipsparse</li>
    <li>rocmPackages.llvm.mlir</li>
    <li>rocmPackages.llvm.openmp</li>
    <li>rocmPackages.llvm.openmp.doc</li>
    <li>rocmPackages.llvm.openmp.info</li>
    <li>rocmPackages.llvm.openmp.man</li>
    <li>rocmPackages.miopen</li>
    <li>rocmPackages.rccl</li>
    <li>rocmPackages.rocalution</li>
    <li>rocmPackages.rocblas</li>
    <li>rocmPackages.rocdbgapi</li>
    <li>rocmPackages.rocdbgapi.doc</li>
    <li>rocmPackages.rocfft</li>
    <li>rocmPackages.rocm-runtime</li>
    <li>rocmPackages.rocminfo</li>
    <li>rocmPackages.rocmlir</li>
    <li>rocmPackages.rocmlir-rock</li>
    <li>rocmPackages.rocmlir.external</li>
    <li>rocmPackages.rocprim</li>
    <li>rocmPackages.rocprofiler</li>
    <li>rocmPackages.rocr-debug-agent</li>
    <li>rocmPackages.rocrand</li>
    <li>rocmPackages.rocsolver</li>
    <li>rocmPackages.rocsparse</li>
    <li>rocmPackages.rocthrust</li>
    <li>rocmPackages.roctracer</li>
    <li>rocmPackages.rocwmma</li>
    <li>rocmPackages.rpp (rocmPackages.rpp-hip)</li>
    <li>rocmPackages.rpp-cpu</li>
    <li>rocmPackages.rpp-opencl</li>
    <li>rocmPackages.tensile</li>
    <li>rocmPackages.tensile.dist</li>
    <li>zluda</li>
  </ul>
</details>

The sizes are looking good as well:
```
du -h --apparent-size --max-depth 0 */ | grep G
3.7G	rocmPackages.composable_kernel/
1.6G	rocmPackages.miopen/
2.7G	rocmPackages.rocblas/
2.1G	rocmPackages.rocfft/
1.5G	rocmPackages.rocsolver/
1.2G	rocmPackages.rocsparse/
```

GZGavinZhao commented:
> You've already patched rocBLAS to run gfx1010 kernels on gfx1012 hardware, so you can further reduce the size by dropping gfx1012 from that list.

@cgmb Good call, thanks for pointing that out! I added `gfx1012` to enable some [optimization/instructions](https://llvm.org/docs/AMDGPU/AMDGPUAsmGFX1011.html) unavailable in `gfx1010`, but forgot my patch causes `gfx101*` GPUs to run as `gfx1010`, so building for `gfx1012` is indeed useless for now :sweat_smile: 

@mschwaig I've removed `gfx1012` just to save a bit more space. I don't think this requires another `nixpkgs-review`, so I think this is good for merge.

mschwaig reviewed (approved):
I used `git range-diff master 062f308b9601f3fec588c8df0225eff7731f9834 32b9f37d57826364db0bc579ee75fa82af1fd623` to verify the contents of that last force-push and it looks good.

I would prefer if we had more info about what targets are missing in that list and why.
Maybe we should have a comment above that `gpuTarget` simply stating that
these are the defaults with `gfx940` and `gfx941` removed because they are early engineering samples and `gfx1012` removed, because of the patch that makes it unused.

Otherwise it will be difficult to get this list right on future ROCm updates.

GZGavinZhao committed:
rocmPackages.rocblas: add comments about why some GPU archs are not built

Signed-off-by: Gavin Zhao <git@gzgz.dev>

GZGavinZhao commented:
@mschwaig Comments added. Ideally I think we should be using `clr.gpuTargets` in all places, though that would be for another PR. Solus uses `gfx803;gfx900;gfx906;gfx908;gfx90a;gfx1010;gfx1030;gfx1100;gfx1101;gfx1102`, and the only package we had to [patch](https://github.com/getsolus/packages/blob/f9f4ad4f3fdd98a44a466f070d83271900f4cb82/packages/c/composable-kernel/package.yml#L49) was `composable_kernel`.

ulrikstrid commented:
I will take a closer look at this tomorrow and see if we can land this. Thanks a lot for working on this!
